### PR TITLE
Add test cases for multiple mapper errors with stop on error

### PR DESCRIPTION
**Pull request by huntharo**
- Add test cases that demonstrate a single exception being rejected when all mappers fail and stop on error is set

### CODE REVIEW

**Comment by sindresorhus** [comment]
When `stopOnError: true`, it's correct that only the first error is catchable, however, the other errors should be silently caught internally.

File: test.js [hunk 1]
```diff
@@ -133,3 +133,36 @@ test('do not run mapping after stop-on-error happened', async t => {
...
+			mappedValues.push(value);
+			await delay(100);
+			throw new Error(`Oops! ${value}`);
+		},
+		{concurrency: 1}),
+		{message: 'Oops! 1'}
+	);
+	t.deepEqual(mappedValues, [1]);
+});
+
```
Comment (sindresorhus) [inline test.js:168]:
```suggestion
```

**Comment by sindresorhus** [comment]
> With concurrent mappers it seems that exceptions should throw an AggregateError when stop on error is true and more than 1 concurrent mapper throws - the iteration of new input items should stop immediately, but all invoked mappers need to finish and all of their exceptions should be bundled and rejected at once - maybe?

When `stopOnError: true`, you're indicating that you want to stop on the first error. AggregateError would not be correct in this case.

**Comment by sindresorhus** [comment]
> I'm not positive how this should work, but it seems wrong that the default behavior will actually start and finish all of the mappers if one of them throws but it only bubbles up 1 random exception even if all of them throw.

That's how `Promise.all` works too.

> In any case, if you decide to keep the current behavior then I can just change the failing test to capture the current behavior to ensure that these behaviors do not accidentally change in the future.

Yeah, I prefer catching all errors, but only returning the first one.

**Comment by huntharo** [comment]
> > I'm not positive how this should work, but it seems wrong that the default behavior will actually start and finish all of the mappers if one of them throws but it only bubbles up 1 random exception even if all of them throw.
> 
> That's how `Promise.all` works too.

Fair point.

I think a distinction here is that with Promise.all the caller has the array of Promises that they can await individually to catch all the exceptions or wait for completion of each item, while with p-map the caller has no way to wait for all created Promises to complete before proceeding as the Promises were not created by the caller and are not exposed to the caller.

A point I'm concerned about here is:

- 1k mappers are started
- 1 mapper fails quickly and throws
- the consumer of p-map has a retry loop
- the retry cannot wait for all the 999 mappers that are still running to complete - thus the retry loop starts before the prior run has finished deterministically
- the consumer ends p with 1,999 mappers running because they have no way to wait until all the previously started mappers are completed before proceeding with their retry

To maintain consistency with Promise.all we'd need to either put a big caveat on `stopOnError: true` indicating that it really shouldn't be used if you need to wait for any mappers to stop before proceeding after one item throws, OR, we could expose an array of mapper promises for the caller to await or ignore at their own risk.  Thoughts on that?

> > In any case, if you decide to keep the current behavior then I can just change the failing test to capture the current behavior to ensure that these behaviors do not accidentally change in the future.
> 
> Yeah, I prefer catching all errors, but only returning the first one.

Ok, let me know what you think of either a doc update with a caveat or exposing an array of mappers that are running when throwing due to `stopOnError: true`.

**Comment by sindresorhus** [comment]
Your issue seems to be more about async/await & promises not having a good cancellation story though, but that's a flaw in the language, not just this package. I do want to support AbortController at some point, while far from ideal, get's us closer: https://github.com/sindresorhus/p-map/issues/51

**Comment by sindresorhus** [comment]
> To maintain consistency with Promise.all we'd need to either put a big caveat on stopOnError: true indicating that it really shouldn't be used if you need to wait for any mappers to stop before proceeding after one item throws, OR, we could expose an array of mapper promises for the caller to await or ignore at their own risk. Thoughts on that?

I think the correct approach is to do https://github.com/sindresorhus/p-map/issues/51 and recommend supporting cancellation in the async tasks they do in the mapper.

> I think stopOnError: true would be most useful if it only threw after all already-started mappers were completed

IMHO, that would be very unexpected behavior and could cause so many other problems. Errors should not be delayed by default. For example, what if one of the mappers it has to wait on takes a very long time to complete, and in the meantime, some other error is thrown somewhere else in the app, and this error is then forever hidden.

**Comment by huntharo** [comment]
> > To maintain consistency with Promise.all we'd need to either put a big caveat on stopOnError: true indicating that it really shouldn't be used if you need to wait for any mappers to stop before proceeding after one item throws, OR, we could expose an array of mapper promises for the caller to await or ignore at their own risk. Thoughts on that?
> 
> I think the correct approach is to do #51 and recommend supporting cancellation in the async tasks they do in the mapper.

Supporting cancellation of the mapper tasks would be helpful here, but it still wouldn't allow deterministically waiting until all of the mappers have completed or exited, right?  So a programmer still could not write code that said "what for that failure to completely stop, then do this".

Perhaps we need to borrow from `p-queue` and have an `.isIdle()` function that returns a promise that has the behavior I'm looking for: all mapper promises have completed (either from cancelling or from completing).

- In the nominal case, with `stopOnError: true`, with no errors the `pMap()` promise would resolve and `.isIdle()` would also immediately resolve if called afterwards
- In the case with an error on the 1st mapper, with `stopOnError: true`, within 1 second and other mappers run for 30 more minutes
  - `pMap()` would reject in 1 second, and `.isIdle()` would resolve in 30 minutes if the mappers did not support cancellation
  - `.isIdle()` would resolve in less than 30 minutes if the mappers did support cancellation
- When `stopOnError: false`, `.isIdle()` could throw since the behavior would be overlapping with the `pMap()` promise
- It could be implemented so that `.isIdle()` would reject with an AggregateError of all exceptions that happened *other than* the first one that caused the `pMap()` rejection

**Comment by sindresorhus** [comment]
> but it still wouldn't allow deterministically waiting until all of the mappers have completed or exited

`{stopOnError: false}`?

**Comment by huntharo** [comment]
> > but it still wouldn't allow deterministically waiting until all of the mappers have completed or exited
> 
> `{stopOnError: false}`?

Yes that works.  Would you accept a documentation caveat being added to the `stopOnError` docs to point out that already-started mappers will continue to run after the first error is returned and that with the default infinite concurrency that essentially all mappers for the entire collection will have already been started when the first error is thrown?

**Comment by sindresorhus** [comment]
> Yes that works. Would you accept a documentation caveat being added to the stopOnError docs to point out that already-started mappers will continue to run after the first error is returned and that with the default infinite concurrency that essentially all mappers for the entire collection will have already been started when the first error is thrown?

Sure, but try to keep it succinct. It would be nice if it also mentioned https://github.com/sindresorhus/p-map/issues/51, so people would be incentivized to contribute towards that feature.

**Comment by huntharo** [comment]
> > Yes that works. Would you accept a documentation caveat being added to the stopOnError docs to point out that already-started mappers will continue to run after the first error is returned and that with the default infinite concurrency that essentially all mappers for the entire collection will have already been started when the first error is thrown?
> 
> Sure, but try to keep it succinct. It would be nice if it also mentioned #51, so people would be incentivized to contribute towards that feature.

I added a doc note... I *think* this one is ready?

File: readme.md [hunk 2]
```diff
@@ -70,6 +70,8 @@ Number of concurrently pending promises returned by `mapper`.
 Type: `boolean`\
 Default: `true`
 
+When set to `true`, the first mapper rejection will be rejected back to the consumer.  Caveat: any already-started async mappers will continue to run until they resolve or reject but they cannot be awaited as their promises are not exposed; in the case of infinite concurrency with sync iterables *all* mappers will be started on startup and will continue after a rejection.  [Issue #51](issues/51) can be implemented to address additional exception and abort use cases.
```
Comment (sindresorhus) [inline readme.md:73]: This is a bit verbose.

And you also need to update index.d.ts.

Comment (sindresorhus) [inline readme.md:73]:
```suggestion
When set to `true`, the first mapper rejection will be rejected back to the consumer. Caveat: any already-started async mappers will continue to run until they resolve or reject but they cannot be awaited as their promises are not exposed; in the case of infinite concurrency with sync iterables *all* mappers will be started on startup and will continue after a rejection. [Issue #51](issues/51) can be implemented to address additional exception and abort use cases.
```

File: test.js [hunk 3]
```diff
@@ -225,3 +225,35 @@ test('catches exception from source iterator - 2nd item after 1st item mapper th
...
+	const mappedValues = [];
+	await t.throwsAsync(
+		pMap(input, async value => {
+			mappedValues.push(value);
+			await delay(100);
+			throw new Error(`Oops! ${value}`);
+		}),
+		{message: 'Oops! 1'}
+	);
+	// Note: all 3 mappers get invoked, all 3 throw, even with stop on error this
```
Comment (sindresorhus) [inline test.js:240]:
```suggestion
	// Note: All 3 mappers get invoked, all 3 throw, even with `{stopOnError: true}` this
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: readme.md (hunk 2 above)
Comment (huntharo) [reply readme.md]: done

File: test.js (hunk 3 above)
Comment (huntharo) [reply test.js]: done

**Comment by sindresorhus** [comment]
Thanks :)

**Pull request merged**